src/ol/interaction/Snap.js

```diff
@@ -288,25 +304,55 @@ class Snap extends PointerInteraction {
     return features;
   }
 
+  /**
+   * Checks if two snap data sets are equal.
+   * Compares the Y-coordinate of the `vertex` and the feature ID.
+   *
+   * @param {SnappedInfo} data1 The first snap data set.
+   * @param {SnappedInfo} data2 The second snap data set.
+   * @return {boolean} `true` if the data sets are equal, otherwise `false`.
+   * @private
+   */
+  areSnapDataEqual_(data1, data2) {
+    return (
+      data1.vertex[1] === data2.vertex[1] &&
+      data1.feature?.getId() === data2.feature?.getId()
+    );
+  }
+
   /**
    * @param {import("../MapBrowserEvent.js").default} evt Map browser event.
    * @return {boolean} `false` to stop event propagation.
    * @api
    */
   handleEvent(evt) {
     const result = this.snapTo(evt.pixel, evt.coordinate, evt.map);
+
     if (result) {
       evt.coordinate = result.vertex.slice(0, 2);
       evt.pixel = result.vertexPixel;
-      this.dispatchEvent(
-        new SnapEvent(SnapEventType.SNAP, {
+
+      // Dispatch UNSNAP event if already snapped
+      if (this.snapped_ && !this.areSnapDataEqual_(this.snapped_, result)) {
+        this.dispatchEvent(new SnapEvent(SnapEventType.UNSNAP, this.snapped_));
+      }
+
+      if (!this.snapped_ || !this.areSnapDataEqual_(this.snapped_, result)) {
```

Comment:
Instead of calling `areSnapDataEqual_()` again here, you could store it in a variable above.